File: 2018/05-polymer-reactions/go/star2.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strings"
)

const caseDifference = 32

func react(polymer string) string {
	var stack []rune
	for _, unit := range polymer {
		var previous rune
		if len(stack) > 0 {
			previous = stack[len(stack)-1]
		}

		charDiff := math.Abs(float64(unit - previous))
		if charDiff == caseDifference {
			// Pop previous
			stack = stack[:len(stack)-1]
			continue
		}

		stack = append(stack, unit)
	}

	return string(stack)
}

func improve(polymer string, char rune) string {
	upperChar := char - caseDifference

	improved := polymer
	improved = strings.Replace(improved, string(char), "", -1)
	improved = strings.Replace(improved, string(upperChar), "", -1)

	return improved
}

func main() {
	input, _ := ioutil.ReadFile("input")
	polymer := strings.TrimSpace(string(input))

	minLength := 99999.0
	for _, char := range "qwertyuiopasdfghjklzxcvbnm" {
		improvedPolymer := improve(polymer, char)
		length := len(react(improvedPolymer))
		minLength = math.Min(minLength, float64(length))
	}
	fmt.Println(minLength)
}

File: 2018/05-polymer-reactions/rust/star2.rs
use std::cmp::min;

const INPUT: &str = include_str!("../input");

fn reacts(c1: char, c2: char) -> bool {
    return (c1 != c2) && (c1.to_uppercase().to_string() == c2.to_uppercase().to_string());
}

fn fully_react(polymer: &str, remove: char) -> usize {
    let mut stack = Vec::new();
    for c in polymer.chars() {
        if c.to_uppercase().to_string() == remove.to_uppercase().to_string() {
            continue;
        }

        if stack.is_empty() {
            stack.push(c);
            continue;
        }

        if reacts(c, stack[stack.len() - 1]) {
            stack.pop();
        } else {
            stack.push(c);
        }
    }

    stack.len()
}

fn main() {
    let polymer = INPUT.trim();

    let mut min_length = 1_000_000;
    for c in "qwertyuiopasdfghjklzxcvbnm".chars() {
        min_length = min(min_length, fully_react(polymer, c));
    }

    println!("{}", min_length);
}
